File: bundler-compiler/src/main/java/com/hendraanggrian/bundler/compiler/UtilsSpec.java
package com.hendraanggrian.bundler.compiler;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.squareup.javapoet.CodeBlock;
import com.squareup.javapoet.MethodSpec;
import com.squareup.javapoet.ParameterSpec;
import com.squareup.javapoet.TypeSpec;
import com.squareup.javapoet.TypeVariableName;

import java.util.Arrays;

import javax.lang.model.element.Modifier;

/**
 * @author Hendra Anggrian (hendraanggrian@gmail.com)
 */
final class UtilsSpec extends Spec {

    @NonNull private final TypeSpec.Builder typeSpec;
    @NonNull private final MethodSpec.Builder methodSpecGet;
    @NonNull private final MethodSpec.Builder methodSpecPut;

    UtilsSpec() {
        TypeVariableName generic = TypeVariableName.get("T");
        Iterable<ParameterSpec> parameterSpecs = Arrays.asList(
                ParameterSpec.builder(CLASS_BUNDLE, "source")
                        .addAnnotation(NonNull.class)
                        .build(),
                ParameterSpec.builder(String.class, "key")
                        .addAnnotation(NonNull.class)
                        .build()
        );
        this.typeSpec = TypeSpec.classBuilder(CLASS_BUNDLER_UTILS.simpleName())
                .addModifiers(Modifier.PUBLIC, Modifier.FINAL);
        this.methodSpecGet = MethodSpec.methodBuilder("getParceler")
                .addTypeVariable(generic)
                .addModifiers(Modifier.PUBLIC, Modifier.STATIC, Modifier.FINAL)
                .addParameters(parameterSpecs)
                .addParameter(ParameterSpec.builder(generic, "defaultValue")
                        .addAnnotation(Nullable.class)
                        .build())
                .addCode(CodeBlock.of("if (source.containsKey(key))\n" +
                        "return $T.unwrap(source.getParcelable(key));\n", CLASS_PARCELS))
                .addStatement("return defaultValue")
                .returns(generic);
        this.methodSpecPut = MethodSpec.methodBuilder("putParceler")
                .addTypeVariable(generic)
                .addModifiers(Modifier.PUBLIC, Modifier.STATIC, Modifier.FINAL)
                .addParameters(parameterSpecs)
                .addParameter(ParameterSpec.builder(generic, "value")
                        .addAnnotation(NonNull.class)
                        .build())
                .addStatement("source.putParcelable(key, $T.wrap(value))", CLASS_PARCELS);
    }

    @NonNull
    @Override
    String packageName() {
        return "com.hendraanggrian.bundler";
    }

    @NonNull
    @Override
    TypeSpec.Builder typeSpec() {
        return typeSpec;
    }

    @NonNull
    @Override
    Iterable<MethodSpec.Builder> methodSpecs() {
        return Arrays.asList(methodSpecGet, methodSpecPut);
    }
}

File: bundler-compiler/src/main/kotlin/com/hendraanggrian/bundler/compiler/UtilsSpec.kt
package com.hendraanggrian.bundler.compiler

import android.os.Bundle
import android.support.annotation.NonNull
import android.support.annotation.Nullable
import com.squareup.javapoet.*
import java.util.*
import javax.lang.model.element.Modifier.*

internal class UtilsSpec : Spec {

    override val typeSpec: TypeSpec.Builder
    private val methodSpecGet: MethodSpec.Builder
    private val methodSpecPut: MethodSpec.Builder

    init {
        val generic = TypeVariableName.get(GENERIC)
        val parameterSpecs = Arrays.asList<ParameterSpec>(
                ParameterSpec.builder(Bundle::class.java, SOURCE)
                        .addAnnotation(NonNull::class.java)
                        .build(),
                ParameterSpec.builder(String::class.java, KEY)
                        .addAnnotation(NonNull::class.java)
                        .build()
        )
        this.typeSpec = TypeSpec.classBuilder(TYPE_BUNDLER_UTILS.simpleName())
                .addModifiers(PUBLIC, FINAL)
        this.methodSpecGet = MethodSpec.methodBuilder("getParceler")
                .addTypeVariable(generic)
                .addModifiers(PUBLIC, STATIC, FINAL)
                .addParameters(parameterSpecs)
                .addParameter(ParameterSpec.builder(generic, DEFAULT_VALUE)
                        .addAnnotation(Nullable::class.java)
                        .build())
                .addCode(CodeBlock.of("if ($SOURCE.containsKey($KEY))\n" +
                        "return \$$GENERIC.unwrap($SOURCE.getParcelable($KEY));\n", TYPE_PARCELS))
                .addStatement("return $DEFAULT_VALUE")
                .returns(generic)
        this.methodSpecPut = MethodSpec.methodBuilder("putParceler")
                .addTypeVariable(generic)
                .addModifiers(PUBLIC, STATIC, FINAL)
                .addParameters(parameterSpecs)
                .addParameter(ParameterSpec.builder(generic, VALUE)
                        .addAnnotation(NonNull::class.java)
                        .build())
                .addStatement("$SOURCE.putParcelable($KEY, \$$GENERIC.wrap($VALUE))", TYPE_PARCELS)
    }

    override val packageName: String get() = "com.hendraanggrian.bundler"

    override val methodSpecs: Iterable<MethodSpec.Builder> get() = listOf(methodSpecGet, methodSpecPut)
}